File: Week 6/PingClient.java
import java.net.*;
import java.io.IOException;

public class PingClient implements Runnable {
	private static final int TIMEOUT = 1000; // milliseconds
	private static final int DELAY = 1000; // milliseconds
	private static final int PACKETS = 10;

	public static void main(String[] args) {
		String host = args[0];
		int port = Integer.parseInt(args[1]);

		Runnable c = new PingClient(host, port);
		c.run();
	}

	private String hostname;
	private InetAddress host;
	private int port;
	private DatagramSocket socket;
	private int packets;

	public PingClient(String hostname, int port) {
		this.hostname = hostname;
		this.port = port;
		packets = PACKETS;
	}

	public void run() {
		if (host == null) {
			try {
				host = InetAddress.getByName(hostname);
			} catch (UnknownHostException e) {
				System.err.println(hostname+": Host name lookup failure");
				return;
			}
		}
		final String sendString = "Ping request";
		final String hostSig = hostname+" ("+host+")";
		System.out.println("PING "+hostSig+" "+sendString.length()+" characters of data.");

		try {
			socket = new DatagramSocket();
			socket.setSoTimeout(TIMEOUT);
		} catch (SocketException e) {
			e.printStackTrace();
			return;
		}

		int seq = 1;

		while (packets-- > 0) {
			byte[] sendData = sendString.getBytes();
			DatagramPacket sendPacket = new DatagramPacket(sendData, sendData.length, host, port);
			try {
				socket.send(sendPacket);
			} catch (IOException e) {
				System.err.println("In sending ping request to "+hostSig+": "+e);
			}
			
			byte[] receiveData = new byte[1024];
			DatagramPacket receivePacket = new DatagramPacket(receiveData, receiveData.length);
			try {
				socket.receive(receivePacket);
				System.out.println(receiveData.length+" bytes from "+hostSig+": icmp_seq="+seq+"");
			} catch (SocketTimeoutException e) {
				System.out.println("Timeout from "+hostSig);
			} catch (IOException e) {
				System.err.println("In receiving ping response from "+hostSig+": "+e);
			}
			++seq;
			try {
				Thread.sleep(DELAY);
			} catch (InterruptedException e) {
				System.err.println("Grr, my beauty sleep!");
			}
		}
	}
}

File: Week 6/pingclient.js
var TIMEOUT = 1000, // milliseconds
    DELAY = 1000, // milliseconds
		PACKETS = 10,
    dgram = require('dgram'),
    dns = require('dns');

if (process.argv.length < 4) {
	console.log("Required arguments: host port");
	process.exit(1);
}

var hostname = process.argv[2],
    port = parseInt(process.argv[3], 10),
		address;

var client = dgram.createSocket('udp4');

var sent = 0, timeouttimer = null;

function generateMessage() {
	return "message";
}

function sendpacket() {
	var msg = generateMessage();
	console.log("Sending "+JSON.stringify(msg)+" to "+address+":"+port);
	client.send(new Buffer(msg), 0, msg.length, port, address);
	++sent;
	if (sent >= PACKETS) {return;}
	timeouttimer = setTimeout(handletimeout, TIMEOUT);
}

function handletimeout() {
	console.log("Timeout from server");
	timeouttimer = null;
	setTimeout(sendpacket, DELAY);
}

client.on('message', function (msg, rinfo) {
	if (timeouttimer !== null) {
		clearTimeout(timeouttimer);
		timeouttimer = null;
	}
	console.log("Received reply from server");
	setTimeout(sendpacket, DELAY);
});

console.log("Looking up "+hostname);
dns.lookup(hostname, null, function (err, addresses) {
	if (err) {throw err;}
	address = addresses;
	console.log(hostname+" = "+address);
	sendpacket();
});
